### PR TITLE
Added lumberjack library for log rolling

### PR DESCRIPTION
**Pull request by evermax**
Hi!
I added the lumberjack library for logs rolling. I did some testing locally, it works like a charme.
I put some default values in it that seemed right, let me know what you think about it.

### CODE REVIEW

**Comment by acmacalister** [comment]
Hey @evermax. Thanks for working on this. I tried to get started on log rolling a few weeks, but just got caught up with my work and school nonsense :smile: . So for what it is worth, I would like to throw my two cents out there.

1. I think instead of repeating the code in the log and errors middleware it would be better to make a helper middleware that the logs and error middleware could import and use. That would save on having to duplicate log management across the two middlewares.

2. Do we want to use a third party library for log rolling? Go's stdlib APIs make log rolling pretty easy. Lumberjack doesn't currently seem to support gzipping the old files, which I think would be a win, regardless if we added it lumberjack or roll it ourselves.

3. I think in the helper middleware we should make a type/struct for configuring log rolling values, so we could allow users to configure the logging policy from the Caddyfile.

Let me know what you think. I should have some free time the week after next and would be happy to help if you need it. Thanks again for starting on this. Can't wait to see this get added!

**Comment by mholt** [comment]
@acmacalister Sorry to jump the gun on this with Maxime - I know you were going to look into log rolling but I know how that burden feels amidst school and work. Since you started on it I really value your feedback as well.

Before this gets merged in, we'll need to make it customizable in the Caddyfile. I agree with Austin that the code could be shared for parsing the Caddyfile, to ensure the syntax is the same.

As for the need for a third-party library, I figure we might as well take advantage of code that's already out there and under test where we need it. Even if we end up forking for some reason in the future we have a good start on a solid implementation.

@evermax Feel free to coordinate with @acmacalister if you'd like help with any of this, I'll also be able to chime in here or on Slack (though my availability will be severely limited this and next week).

**Comment by evermax** [comment]
I agree with @mholt that it is easier for maintenance to use an existing library. I will have a look at the helper to avoid duplicating log management tomorrow, and also how to make the user able to provide his options for the log rolling. If you have any more input on that @acmacalister, let me know!
We can talk about it more in details over Slack or here.

**Comment by evermax** [comment]
I am a bit struggling with the idea of sharing the same parser on `log` and `errors` for customised values.
The directives don't have the same format in the Caddyfile. @mholt what should be the structure here?
For example in the `log` directive, we have 3 different ways of writing the configuration in the Caddy file: which is/are the one that should allow the add parameters for the log rolling?
Should we create the ability to have a block, in which we add a parameter rolling like this?
```
log file path {
   rolling 20 10 etc.
}
```
In that case, perhaps it would be easier to have something more consistant with the `errors` directive configuration approach.
What do you think?

**Comment by mholt** [comment]
Initial ideas:

```
log {
	rotate {
		size 100  # MB
		age 10  # days
		max 5  # number of old backups
	}
}

errors {
	log errors.log {
		size 100
		age 10
		max 5
	}
}
```

The code which parses the "rotate" block (between the curly braces, not including the word rotate because the errors directive does not use it) could be shared.

What do you think?

**Comment by evermax** [comment]
It seems nice, I will try this out.
It is simple and yet explicit enough, I think!

**Comment by mholt** [comment]
Maybe rename "max" to "keep"

File: config/setup/errors.go [hunk 1]
```diff
@@ -71,6 +81,16 @@ func errorsParse(c *Controller) (*errors.ErrorHandler, error) {
 
 			if what == "log" {
 				handler.LogFile = where
+				if c.NextArg() {
+					if c.Val() == "{" {
+						c.IncrNest()
```
Comment (mholt) [inline config/setup/errors.go]: I think you did a very good job here of handling the unusual, nested block. I can see what you mean about how it's not pretty. For now I'll take it and if another nested block situation comes up, I'll put some more effort into improving on this. It's a great start!

Also the tests are great!

**Comment by mholt** [comment]
@evermax Excellent, this is great! Thanks for working on it. Sorry the nested block was kind of a mess. But you handled it well.

@acmacalister I'm gonna merge this in. What do you think?

**Comment by acmacalister** [comment]
LGTM. Great Job!

**Comment by evermax** [comment]
Cool, thanks guys for your review and guidance on that, happy to help!

**Comment by mholt** [comment]
@acmacalister Re: your comment about compressing old files; see https://github.com/natefinch/lumberjack/issues/13

**Comment by natefinch** [comment]
FWIW, lumberjack has been in use as a part of [Juju](http://github.com/juju/juju) for over a year, so it definitely has had a lot of burn-in (and we've fixed at least one bug in it from that time period).  Definitely, I'd rather see the code reused rather than duplicate implementations pop up - not out of ego, but just to consolidate the community's efforts.

I'm psyched to see it used by caddy.  I'd love to see some code for compression of backups, I just haven't gotten around to it, and have been super busy lately.  Feel free to propose something.  I made a sketch of the type of thing I'd write given the time to do so, but other proposals are welcome.  There's some trickiness with timing and cleaning up old backups as well as what happens if you stop the application and restart it before the compression had finished.

**Comment by acmacalister** [comment]
@natefinch I agree, in most contexts duplicating efforts sucks. I wasn't sure about the policy on third party libraries within Caddy. I actually forked lumberjack a couple weeks ago to start adding compression on backups, but that is about as far as I got. :smile: I will share my idea on compressed backups on the issue Matt opened above.

**Pull request merged**